# Άσκηση 33-35



In [4]:
import random
random.seed(2016)  # for reproducibility
import statistics as stat
from simple_markov_chain_lib import markov_chain

p = 1/6

# A dictionary for the initial distibution. 
# We prescribe the initial distribution
init_probs = {0: 1.0} 
 
# A dictionary for the transition probability  matrix. 
# Every state-key corresponds to a list with tuples of (Next_State,Probability) 
markov_table = {
    0: {1: 1.},
    1: {1: 2/3, 2: 1/3},
    2: {0: p, 1: 1-p}
}
 
# Ok... we are ready now
# Let's construct a Markov Chain. So let's call the constructor
mc = markov_chain(markov_table, init_probs)

phat = []
M = 50

for j in range(M):
    ## Experiment parameters
    N = 200 # or N = 20000   # number of samples
    steps = 40   # the target time
    counter = 0  # to count the number of times the event {X_40  = 0} occurs

    ## Simulation
    for i in range(N):
        mc.start()  # new experiment
        for j in range(steps):  mc.move()
        if mc.running_state == 0:  counter += 1

    phat.append(counter / N)
    
print(
    """ 
    The sample mean is {0:.5f} and the sample variance is {1:.5f}
    """.format(stat.mean(phat), stat.variance(phat))
)


 
    The sample mean is 0.03870 and the sample variance is 0.00020
    


## Απαντήσεις

1. Για $N = 200$ η μέση εκτίμηση της $\mathbb{P}_0[X_{40}=0]$ είναι $0.03870$, ενώ για $N = 20000$ είναι $0.04012$.
2. Η θεωρητική τιμή που υπολογίσαμε ήταν $\mathbb{P}_0[X_{40}=0] = 0.04$, επομένως τα αποτελέσματα είναι αρκετά κοντά.
3. Για $N = 200$ η δειγματική διασπορά των $50$ εκτιμήσεων είναι $0.00020$, ενώ για $N = 20000$ είναι $0.00000$.
4. Όχι, καθώς ο αριθμός των φορών που προσομοιάζεται η αλυσίδα είναι πολύ μεγάλος και στις $2$ περιπτώσεις, γεγονός που καθιστά την αρχική κατανομή της αλυσίδας αδιάφορη. Αυτό επαληθεύεται εύκολα και αν θέσουμε init_probs = {2: 1.0} και τρέξουμε ξανά τον παραπάνω κώδικα. 


# Άσκηση 38

In [5]:
import random
random.seed(2016)  # for reproducibility
import statistics as stat
from simple_markov_chain_lib import markov_chain

name1 = "0-0"
name2 = "15-0"
name3 = "0-15"
name4 = "30-0"
name5 = "15-15"
name6 = "0-30"
name7 = "40-0"
name8 = "30-15"
name9 = "15-30"
name10 = "0-40"
name11 = "GameA"
name12 = "40-15"
name13 = "30-30"
name14 = "15-40"
name15 = "GameB"
name16 = "40-30"
name17 = "30-40"
name18 = "Deuce"
name19 = "advA"
name20 = "advB"

p = .6

# A dictionary for the transition probability  matrix. 
# Every state-key corresponds to a list with tuples of (Next_State,Probability) 
markov_table = {
    name1: {name2: p, name3: 1-p},
    name2: {name4: p, name5: 1-p},
    name3: {name5: p, name6: 1-p},
    name4: {name7: p, name8: 1-p},
    name5: {name8: p, name9: 1-p},
    name6: {name9: p, name10: 1-p},
    name7: {name11: p, name12: 1-p},
    name8: {name12: p, name13: 1-p},
    name9: {name13: p, name14: 1-p},
    name10: {name14: p, name15: 1-p},
    name11: {name11: 1},
    name12: {name11: p, name16: 1-p},
    name13: {name16: p, name17: 1-p},
    name14: {name17: p, name15:1-p},
    name15: {name15: 1},
    name16: {name11: p, name18: 1-p},
    name17: {name18: p, name15: 1-p},
    name18: {name19: p, name20:1-p},
    name19: {name11: p, name18:1-p},
    name20: {name18: p, name15:1-p}
}
 
# A dictionary for the initial distribution. 
# We prescribe the initial distribution
init_probs = {name1: 1.0} 


# Ok... we are ready now
# Let's construct a Markov Chain. So let's call the constructor
mc = markov_chain(markov_table, init_probs)
## Experiment parameters
N = 1000000  # number of samples
counter = 0  # to count the number of times the event occurs

## Simulation
for i in range(N):
    mc.start()  # new experiment
    while (mc.running_state != name11 and mc.running_state != name15):  mc.move()  
    if mc.running_state == name11:  counter += 1
            
phat = (counter / N)
    
print(
    """
    We executed {0} times the markov chain
    and player A won {1} times.
    So we estimate the probability of him to win to be {2}
    """.format(N, counter, phat)
)



    We executed 1000000 times the markov chain
    and player A won 735455 times.
    So we estimate the probability of him to win to be 0.735455
    
